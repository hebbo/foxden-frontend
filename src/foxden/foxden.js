// @flow
import React, { Component } from "react";
import ReactDOM from "react-dom";

type Post = {
  id: string,
  userId: string,
  imageUrl: string,
  description: string
};

type User = {
  id: string,
  username: string
};

type FoxData = {
  posts: Array<Post>,
  users: Array<User>
};

type Props = { data: FoxData };

function getUserByID(users: Array<User>, id: string): ?User {
  return users.find(user => user.id === id) || null;
}

class App extends Component<Props, {}> {
  render() {
    return this.props.data.posts.map(this_post => (
      <PostComponent
        key={this_post.id}
        post={this_post}
        users={this.props.data.users}
        getUserFunction={getUserByID}
      />
    ));
  }
}

function PostComponent(props: {
  post: Post,
  users: Array<User>,
  getUserFunction: (Array<User>, string) => ?User
}) {
  let { post } = props;
  let post_user = props.getUserFunction(props.users, post.userId);
  return (
    <div>
      <img src={post.imageUrl} alt={post.description} />
      <p>{post_user ? post_user.username : "Anonymus"}</p>
    </div>
  );
}

// This just exports the variables so we can test them with jest
export { PostComponent, App, getUserByID };
