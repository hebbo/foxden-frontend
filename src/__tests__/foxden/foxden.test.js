/**
 * @jest-environment node
 */
// @flow
import React from "react";
import Adapter from "enzyme-adapter-react-16";
import Enzyme, { shallow } from "enzyme";

import { shallowToJson } from "enzyme-to-json";

import { PostComponent, App, getUserByID } from "../../foxden/foxden";

Enzyme.configure({ adapter: new Adapter() });

const foxData: Object = JSON.parse(
  '{"posts":[{"id":"6cbe1c74","userId":"02","imageUrl":"https://randomfox.ca/images/10.jpg","description":"Lonefloof"}],"users":[{"id":"02","username":"XXX"}]}'
);

describe("App", () => {
  it("should render correctly", () => {
    const output = shallow(<App data={foxData} />);
    expect(shallowToJson(output)).toMatchSnapshot();
  });
});

describe("PostComponent", () => {
  it("should render correctly", () => {
    const output = shallow(
      <PostComponent
        key={foxData.posts[0].id}
        post={foxData.posts[0]}
        users={foxData.users}
        getUserFunction={getUserByID}
      />
    );
    expect(shallowToJson(output)).toMatchSnapshot();
  });
});

describe("getUserByID", () => {
  test("extract user correctly", () => {
    let actual = { id: "02", username: "XXX" };
    expect(getUserByID(foxData.users, "02")).toEqual(actual);
  });
});
