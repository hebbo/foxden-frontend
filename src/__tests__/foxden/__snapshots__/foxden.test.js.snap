// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
Array [
  <PostComponent
    getUserFunction={[Function]}
    key="6cbe1c74"
    post={
      Object {
        "description": "Lonefloof",
        "id": "6cbe1c74",
        "imageUrl": "https://randomfox.ca/images/10.jpg",
        "userId": "02",
      }
    }
    users={
      Array [
        Object {
          "id": "02",
          "username": "XXX",
        },
      ]
    }
  />,
]
`;

exports[`PostComponent should render correctly 1`] = `
<div>
  <img
    alt="Lonefloof"
    src="https://randomfox.ca/images/10.jpg"
  />
  <p>
    XXX
  </p>
</div>
`;
